File: src/claw_mem/storage/ground_truth.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
GroundTruthStore - Complete raw conversation preservation.

Stores full conversation transcripts per session for traceability
and deep analysis. Complements Episodic Storage (structured memories)
by preserving the original dialogue in its entirety.
"""

import json
import os
import time
import uuid
from pathlib import Path
from typing import Dict, List, Optional
from dataclasses import dataclass, field


@dataclass
class GroundTruthRecord:
    """A single raw conversation record."""

    record_id: str
    session_id: str
    messages: List[Dict]
    timestamp: float
    metadata: Dict = field(default_factory=dict)

    def to_dict(self) -> dict:
        return {
            "record_id": self.record_id,
            "session_id": self.session_id,
            "messages": self.messages,
            "timestamp": self.timestamp,
            "metadata": self.metadata,
        }

    @classmethod
    def from_dict(cls, d: dict) -> "GroundTruthRecord":
        return cls(
            record_id=d["record_id"],
            session_id=d["session_id"],
            messages=d.get("messages", []),
            timestamp=d.get("timestamp", 0.0),
            metadata=d.get("metadata", {}),
        )


class GroundTruthStore:
    """Preserves full raw conversation transcripts.

    Storage layout:
        ~/.claw-mem/ground_truth/
            session_{id}.json   # One JSON file per session

    Design: append-only files; one JSON array per session file.
    """

    def __init__(self, workspace: str = None):
        if workspace is None:
            workspace = os.path.expanduser("~/.claw-mem")
        self._base_dir = Path(workspace) / "ground_truth"
        self._base_dir.mkdir(parents=True, exist_ok=True)

    def _session_path(self, session_id: str) -> Path:
        safe_id = session_id.replace("/", "_").replace(":", "_")
        return self._base_dir / f"session_{safe_id}.json"

    def store_turn(self, session_id: str, messages: List[Dict],
                   metadata: Optional[Dict] = None) -> str:
        """Store one or more conversation turns.

        Args:
            session_id: OpenClaw session identifier.
            messages: List of message dicts with 'role' and 'content'.
            metadata: Optional metadata for this batch.

        Returns:
            record_id (format: gt_<hex16>).
        """
        record = GroundTruthRecord(
            record_id=f"gt_{uuid.uuid4().hex[:16]}",
            session_id=session_id,
            messages=messages,
            timestamp=time.time(),
            metadata=metadata or {},
        )
        filepath = self._session_path(session_id)
        session_data = self._load_session_file(filepath)
        session_data.append(record.to_dict())
        self._save_session_file(filepath, session_data)
        return record.record_id

    def store_session(self, session_id: str,
                      all_messages: List[Dict],
                      metadata: Optional[Dict] = None) -> str:
        """Store an entire session at once (called from agent_end)."""
        return self.store_turn(session_id, all_messages, metadata)

    def get_session(self, session_id: str) -> List[Dict]:
        """Get all records for a session."""
        return self._load_session_file(self._session_path(session_id))

    def search(self, session_id: Optional[str] = None,
               keyword: Optional[str] = None,
               limit: int = 50) -> List[Dict]:
        """Search raw conversations.

        Args:
            session_id: Optional session filter.
            keyword: Optional content keyword filter.
            limit: Maximum results.
        """
        results: List[Dict] = []
        kw_lower = keyword.lower() if keyword else None

        if session_id:
            paths = [self._session_path(session_id)]
        else:
            paths = sorted(
                self._base_dir.glob("session_*.json"),
                key=os.path.getmtime, reverse=True,
            )

        for fp in paths:
            sid = fp.stem.replace("session_", "") if isinstance(fp, Path) else None
            records = self._load_session_file(fp)
            for r in records:
                for m in r.get("messages", []):
                    if kw_lower and kw_lower not in str(m).lower():
                        continue
                    results.append({
                        "session_id": sid or session_id,
                        "message": m,
                        "timestamp": r.get("timestamp", 0),
                    })
                    if len(results) >= limit:
                        return results
        return results[:limit]

    def list_sessions(self) -> List[Dict]:
        """List all stored sessions with metadata."""
        sessions = []
        for f in sorted(
            self._base_dir.glob("session_*.json"),
            key=os.path.getmtime, reverse=True,
        ):
            sessions.append({
                "session_id": f.stem.replace("session_", ""),
                "file_size": f.stat().st_size,
                "last_modified": f.stat().st_mtime,
            })
        return sessions

    def count_records(self) -> int:
        total = 0
        for f in self._base_dir.glob("session_*.json"):
            data = self._load_session_file(f)
            total += len(data)
        return total

    # ── Internal helpers ────────────────────────────────────────────

    def _load_session_file(self, filepath: Path) -> List[Dict]:
        if not filepath.exists():
            return []
        try:
            with open(filepath, 'r', encoding='utf-8') as f:
                return json.load(f)
        except (json.JSONDecodeError, FileNotFoundError):
            return []

    def _save_session_file(self, filepath: Path, data: List[Dict]) -> None:
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(data, f, ensure_ascii=False, indent=2)

File: src/storage/ground_truth.ts
import type { GroundTruthSessionRecord } from "../types.js";
import * as fs from "fs";
import * as path from "path";
import * as os from "os";
import * as crypto from "crypto";

/**
 * GroundTruthStore — preserves full raw conversation transcripts.
 *
 * Storage: `ground_truth/session_{id}.json` — one append-only JSON
 * array per session file. Design matches Python GroundTruthStore.
 */
export class GroundTruthStore {
  private baseDir: string;

  constructor(workspace?: string) {
    this.baseDir = path.join(workspace || path.join(os.homedir(), ".claw-mem"), "ground_truth");
    fs.mkdirSync(this.baseDir, { recursive: true });
  }

  /** Store one or more conversation turns. Returns record_id. */
  storeTurn(
    sessionId: string,
    messages: Array<{ role: string; content: string }>,
    metadata?: Record<string, unknown>
  ): string {
    const recordId = `gt_${crypto.randomUUID().replace(/-/g, "").slice(0, 16)}`;
    const record: GroundTruthSessionRecord = {
      record_id: recordId,
      session_id: sessionId,
      messages,
      timestamp: Date.now() / 1000,
      metadata: metadata || {},
    };
    const filePath = this.sessionPath(sessionId);
    const sessionData = this.loadSessionFile(filePath);
    sessionData.push(record);
    this.saveSessionFile(filePath, sessionData);
    return recordId;
  }

  /** Store an entire session at once. */
  storeSession(
    sessionId: string,
    allMessages: Array<{ role: string; content: string }>,
    metadata?: Record<string, unknown>
  ): string {
    return this.storeTurn(sessionId, allMessages, metadata);
  }

  /** Get all records for a session. */
  getSession(sessionId: string): GroundTruthSessionRecord[] {
    return this.loadSessionFile(this.sessionPath(sessionId));
  }

  /** Search raw conversations by session/keyword. */
  search(
    sessionId?: string,
    keyword?: string,
    limit: number = 50
  ): Array<{ session_id: string; message: unknown; timestamp: number }> {
    const results: Array<{ session_id: string; message: unknown; timestamp: number }> = [];
    const kwLower = keyword?.toLowerCase();
    const paths = sessionId
      ? [this.sessionPath(sessionId)]
      : this.listFiles().map((f) => path.join(this.baseDir, f));
    for (const fp of paths) {
      const sid = path.basename(fp).replace("session_", "").replace(".json", "");
      const records = this.loadSessionFile(fp);
      for (const r of records) {
        for (const m of r.messages) {
          if (kwLower && !JSON.stringify(m).toLowerCase().includes(kwLower)) continue;
          results.push({ session_id: sid, message: m, timestamp: r.timestamp });
          if (results.length >= limit) return results;
        }
      }
    }
    return results.slice(0, limit);
  }

  /** List all stored sessions with metadata. */
  listSessions(): Array<{ sessionId: string; fileSize: number; lastModified: number }> {
    const sessions: Array<{ sessionId: string; fileSize: number; lastModified: number }> = [];
    for (const fname of this.listFiles()) {
      const fp = path.join(this.baseDir, fname);
      try {
        const stat = fs.statSync(fp);
        sessions.push({
          sessionId: fname.replace("session_", "").replace(".json", ""),
          fileSize: stat.size,
          lastModified: stat.mtimeMs,
        });
      } catch { continue; }
    }
    sessions.sort((a, b) => b.lastModified - a.lastModified);
    return sessions;
  }

  /** Count total records across all sessions. */
  countRecords(): number {
    let total = 0;
    for (const fname of this.listFiles()) {
      total += this.loadSessionFile(path.join(this.baseDir, fname)).length;
    }
    return total;
  }

  // ── internal ────────────────────────────────────────────────────

  private sessionPath(sessionId: string): string {
    const safeId = sessionId.replace(/[/:]/g, "_");
    return path.join(this.baseDir, `session_${safeId}.json`);
  }

  private listFiles(): string[] {
    try {
      return fs.readdirSync(this.baseDir).filter((f) => f.startsWith("session_") && f.endsWith(".json"));
    } catch { return []; }
  }

  private loadSessionFile(filePath: string): GroundTruthSessionRecord[] {
    if (!fs.existsSync(filePath)) return [];
    try {
      return JSON.parse(fs.readFileSync(filePath, "utf-8")) as GroundTruthSessionRecord[];
    } catch { return []; }
  }

  private saveSessionFile(filePath: string, data: GroundTruthSessionRecord[]): void {
    const tmpPath = filePath + ".tmp." + Date.now();
    fs.writeFileSync(tmpPath, JSON.stringify(data, null, 2), "utf-8");
    fs.renameSync(tmpPath, filePath);
  }
}
